src/main/java/com/google/devtools/build/lib/analysis/constraints/TopLevelConstraintSemantics.java

```diff
@@ -88,6 +106,54 @@ private MissingEnvironment(Label environment, RemovedEnvironmentCulprit culprit)
     }
   }
 
+  /**
+   * Checks that the all top-level targets are compatible with the target platform.
+   *
+   * <p>If any target doesn't support the target platform it will be either marked as "to be kipped"
+   * or marked as "errored".
+   *
+   * <p>Targets that are incompatible with the target platform and are not explicitly requested on
+   * the command line should be skipped.
+   *
+   * <p>Targets that are incompatible with the target platform and *are* explicitly requested on the
+   * command line are errored. Having one or more errored targets will cause the entire build to
+   * fail with an error message.
+   */
+  public PlatformRestrictionsResult checkPlatformRestrictions(
+      ImmutableList<ConfiguredTarget> topLevelTargets,
+      List<String> requestedTargetPatterns,
+      boolean keepGoing)
+      throws ViewCreationFailedException {
+    ImmutableSet.Builder<ConfiguredTarget> incompatibleTargets = ImmutableSet.builder();
+    ImmutableSet.Builder<ConfiguredTarget> incompatibleButRequestedTargets = ImmutableSet.builder();
+
+    for (ConfiguredTarget target : topLevelTargets) {
+      if (target.getProvider(IncompatiblePlatformProvider.class) != null) {
+        String labelString = target.getLabel().toString();
+        if (requestedTargetPatterns.contains(labelString)) {
```

Comment:
This is buggy. `requestedTargetPatterns` can be something like `:foo` while `labelString` could be `//:foo`.

Need to find a better way to detect this.

Reply:
@gregestren , you wouldn't happen to know of a good way around this, would you?

The `requestedTargetPatterns` are strings equivalent to what's specified on the command line. I think I need to resolve the non-globbing ones to Labels so that the comparison here makes more sense. My thought is to enhance the call chain in `AnalysisPhaseRunner.evaluateTargetPatterns()` somehow to keep track of which targets were explicitly requested. That is leading me on a deep dive. I'm currently looking in `SequentialBuildDriver.evaluate()`. Will resume on this later.

If you have any hints, I'd appreciate it :)

Reply:
Commented above.